nimare/generate.py

```diff
@@ -0,0 +1,290 @@
+    else:
+        if len(n_foci) != n_studies:
+            raise ValueError("n_foci must be the same length as n_studies")
+        n_focis = n_foci
+    # foci_coords are to be generated randomly
+    if foci_coords is None:
+        foci_idxs = rng.choice(range(possible_ijks.shape[0]), max(n_focis), replace=False)
+        # if there are no foci_idxs, give a dummy coordinate (0, 0, 0)
+        ground_truth_foci_ijk = (
+            possible_ijks[foci_idxs] if foci_idxs.size else np.array([(0, 0, 0)])
+        )
+        foci_coords_num = ground_truth_foci_ijk.shape[0]
+    # foci_coords were specified by the caller
+    elif isinstance(foci_coords, list):
+        ground_truth_foci_ijk = [mm2vox(coord, template_img.affine) for coord in foci_coords]
+        foci_coords_num = len(ground_truth_foci_ijk)
+    else:
+        raise ValueError("foci_coords must be a list or None")
+
+    if not foci_weights:
+        foci_weights = [1] * foci_coords_num
+    if not isinstance(foci_weights, list):
+        raise ValueError("foci_weights must be a list")
+    elif len(foci_weights) != foci_coords_num:
+        raise ValueError("foci_weights must be the same length as foci")
+
+    if isinstance(fwhm, float):
+        fwhms = [fwhm] * foci_coords_num
+    elif isinstance(fwhm, list):
+        fwhms = [float(f) for f in fwhm]
+        if len(fwhms) != foci_coords_num:
+            raise ValueError(
+                ("fwhm must be a list the same size as the number of foci or a float")
+            )
+    else:
+        try:
+            fwhms = [float(fwhm)] * foci_coords_num
+        except ValueError:
+            raise ValueError("fwhm must be a float or a list")
+
+    foci_dict = {}
+    # generate study specific foci for each ground truth focus
+
+    kernels = [get_ale_kernel(template_img, fwhm)[1] for fwhm in fwhms]
+    weighted_prob_map = sum(
+        [
+            compute_ma(template_data.shape, np.atleast_2d(ground_truth_focus), kernel) * weight
+            for kernel, ground_truth_focus, weight in zip(
+                kernels, ground_truth_foci_ijk, foci_weights
+            )
+        ]
+    )
+    # handle case when sum is 0 by adding a small amount
+    weighted_prob_map = weighted_prob_map / np.sum(weighted_prob_map) + DEF_TINY
+    weighted_prob_map_ijks = np.argwhere(weighted_prob_map)
+    weighted_prob_vector = weighted_prob_map[np.nonzero(weighted_prob_map)]
+
+    if isinstance(n_noise_foci, int):
+        n_noise_foci_list = [n_noise_foci] * n_studies
+    else:
+        n_noise_foci_list = n_noise_foci
+
+    for f_num, f_noise, study in zip(n_focis, n_noise_foci_list, range(n_studies)):
+        ijk_idxs = rng.choice(
+            weighted_prob_map_ijks.shape[0], f_num, p=weighted_prob_vector, replace=False
+        )
+        foci_ijks = weighted_prob_map_ijks[ijk_idxs]
+
+        if f_noise > 0:
+            weighted_noise_map_ijks = np.delete(weighted_prob_map_ijks, ijk_idxs, axis=0)
```

Comment:
If I understand this correctly, I'm not sure we need/want this... Noise could in theory arise in the same coordinates as signal, so I don't think we should forcibly exclude signal voxels from contention.

Reply:
from our conversation today (10/28/2020), I have this changed.